odoo/fields.py

```diff
@@ -3315,6 +3315,80 @@ def convert_to_onchange(self, value, record, names):
         self._add_display_name(value, record.env)
         return value
 
+    def read(self, records):
+        """Read everything needed in batch for the given records.
+
+        To retrieve relational properties names, or to check their existence,
+        we need to do some SQL queries. To reduce the number of queries when we read
+        in batch, we put in cache everything needed before calling
+        convert_to_record / convert_to_read.
+        """
+        definition_records_map = {
+            record: record[self.definition_record][self.definition_record_field]
+            for record in records
+        }
+
+        # ids per model we need to fetch in batch to put in cache
+        ids_per_model = defaultdict(list)
+
+        cached_values = list(records.env.cache.get_values(records, self))
+
+        for record, values in zip(records, cached_values):
+            definition = definition_records_map.get(record)
+            if not values or not definition:
+                continue
+            for property_definition in definition:
+                comodel = property_definition.get('comodel')
+                type_ = property_definition.get('type')
+                name = property_definition.get('name')
+                if not comodel or type_ not in ('many2one', 'many2many') or name not in values:
+                    continue
+
+                default = property_definition.get('default') or []
+                value = values[name] or []
+
+                if type_ == 'many2one':
+                    default = [default] if default else []
+                    value = [value] if value else []
+
+                ids_per_model[comodel].extend(default + value)
+
+        # check existence and pre-fetch in batch
+        existing_ids_per_model = {}
+        for model, ids in ids_per_model.items():
+            try:
+                rec = records.env[model].browse(ids).exists()
+                existing_ids_per_model[model] = rec.ids
+                rec.mapped('create_uid')  # read a field to pre-fetch the recordset
+            except AccessError:
```

Comment:
You should filter `rec` with ir.rule before. In case, you have a AccessError for the first record but not for others, the cache won't be filled for others ?